### PR TITLE
JAMES-2664 Guice loading for PreDeletionHooks

### PR DESCRIPTION
**Pull request by chibenwa**
I included a little refactoring of MailboxListeners loading that I used as a basis for this work.

### CODE REVIEW

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/PreDeletionHookModule.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+public class PreDeletionHookModule extends AbstractModule {
+    @Override
+    protected void configure() {
+        bind(PreDeletionHookLoader.class).to(PreDeletionHookLoaderImpl.class);
+    }
+
+    @Provides
+    PreDeletionHooksConfiguration providesConfiguration(ConfigurationProvider configurationProvider) throws ConfigurationException {
+        return PreDeletionHooksConfiguration.from(configurationProvider.getConfiguration("listeners"));
```
Comment (hoangdat) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/PreDeletionHookModule.java]: Hook will be config on the same with mailbox listener?

Can we need add doc or template for hook configuration?

Comment (chibenwa) [reply server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/PreDeletionHookModule.java]: Please contribute Guice bindings separately

Comment (chibenwa) [reply server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/PreDeletionHookModule.java]: Hmm I must admit  I did not really wanted to introduce 'yet another' configuration file hence I decided to do that as part of listeners.xml...

Let's do the DOC part a bit later, as documenting the all feature is needed.

File: server/container/guice/mailbox/src/test/java/org/apache/james/modules/mailbox/PreDeletionHooksConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    void fromShouldReturnNoneWhenEmpty() throws Exception {
+        HierarchicalConfiguration configuration = new HierarchicalConfiguration();
+
+        assertThat(PreDeletionHooksConfiguration.from(configuration))
+            .isEqualTo(PreDeletionHooksConfiguration.none());
+    }
+
+    @Test
+    void fromShouldThrowWhenInvalidHookConfiguration() throws Exception {
+        XMLConfiguration configuration = new XMLConfiguration();
```
Comment (aduprat) [inline server/container/guice/mailbox/src/test/java/org/apache/james/modules/mailbox/PreDeletionHooksConfigurationTest.java]: Usually we are using _PropertiesConfiguration_.

Comment (chibenwa) [reply server/container/guice/mailbox/src/test/java/org/apache/james/modules/mailbox/PreDeletionHooksConfigurationTest.java]: Well, `listeners.xml` is in XML

Also, properties are bad at expressing lists of complex elements.

Finally, keeping easy evolution toward a configuration per-hook seems to be an important concern.

All of this tends to promote XML backed hierarchicalConfiguration.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**